smack-integration-test/src/main/java/org/igniterealtime/smack/inttest/util/MultiResultSyncPoint.java

```diff
@@ -0,0 +1,74 @@
+/**
+ *
+ * Copyright 2021 Guus der Kinderen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.igniterealtime.smack.inttest.util;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeoutException;
+
+import org.jivesoftware.smack.util.Objects;
+
+public class MultiResultSyncPoint<R, E extends Exception> {
+
+    private final List<R> results;
+    private E exception;
+    private int expectedResultCount;
+
+    public MultiResultSyncPoint(int expectedResultCount) {
+        this.expectedResultCount = expectedResultCount;
+        this.results = new ArrayList<>(expectedResultCount);
+    }
+
+    public List<R> waitForResults(long timeout) throws E, InterruptedException, TimeoutException {
+        synchronized (this) {
+            if (expectedResultCount <= results.size()) {
+                return results;
+            }
+            if (exception != null) {
+                throw exception;
+            }
+            final long deadline = System.currentTimeMillis() + timeout;
+            while (expectedResultCount > results.size() && exception == null) {
+                final long now = System.currentTimeMillis();
+                if (now >= deadline) break;
+                wait(deadline - now);
+            }
+        }
+        if (expectedResultCount <= results.size()) {
```

Comment:
What duplication are you referring to here? I don't see how we can reduce the check if the expected count matches the result count, as that needs to happen both before and after waiting. (This mimics what's implemented in `ResultSyncPoint`).

Reply:
First, I just noticed that this method directly returns results. I am not sure if this is sensible, given that results could potentially concurrently be modified after it was returned. Therefore I would suggest returning a copy, where the copy is made while the monitor lock is held.

I suggest the following, IMHO more idiomatic, implementation for the method. Also note that it follows a more defensive approach, where the exception is thrown, even if all results are there. If there is a compelling reason to swap this, then I could consider that. (It would also be made configurable)

```java
    public synchronized List<R> waitForResults(long timeout) throws E, InterruptedException, TimeoutException {
        long now = System.currentTimeMillis();
        final long deadline = now + timeout;
        while (results.size <= expectedResultSize && !exception && now < deadline) {
              wait(deadline - now);
              now = System.currentTimeMillis();
        }
         if (now >= deadline) throw new TimeoutException("Timeout waiting " + timeout + " millis");
         if (exception != null) throw exception;
         return results; // TODO: Copy results
    }
```